// sequelize-auto/lib/auto.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SequelizeAuto = void 0;
const lodash_1 = __importDefault(require("lodash"));
const sequelize_1 = require("sequelize");
const auto_builder_1 = require("./auto-builder");
const auto_generator_1 = require("./auto-generator");
const auto_writer_1 = require("./auto-writer");
const dialects_1 = require("./dialects/dialects");
class SequelizeAuto {
    constructor(database, username, password, options) {
        if (options && options.dialect === 'sqlite' && !options.storage && database) {
            options.storage = database;
        }
        if (options && options.dialect === 'mssql') {
            // set defaults for tedious, to silence the warnings
            options.dialectOptions = options.dialectOptions || {};
            options.dialectOptions.options = options.dialectOptions.options || {};
            options.dialectOptions.options.trustServerCertificate = true;
            options.dialectOptions.options.enableArithAbort = true;
            options.dialectOptions.options.validateBulkLoadParameters = true;
        }
        if (database instanceof sequelize_1.Sequelize) {
            this.sequelize = database;
        }
        else {
            this.sequelize = new sequelize_1.Sequelize(database, username, password, options || {});
        }
        this.options = lodash_1.default.extend({
            spaces: true,
            indentation: 2,
            directory: './models',
            additional: {},
            freezeTableName: true,
            host: 'localhost',
            port: this.getDefaultPort(options.dialect),
            closeConnectionAutomatically: true
        }, options || {});
    }
    async run() {
        const td = await this.build();
        const tt = this.generate(td);
        await this.write(tt);
        td.text = tt;
        return td;
    }
    build() {
        const builder = new auto_builder_1.AutoBuilder(this.sequelize, this.options.tables, this.options.skipTables, this.options.schema, this.options.views);
        return builder.build().then(tableData => {
            if (this.options.closeConnectionAutomatically) {
                return this.sequelize.close().then(() => tableData);
            }
            return tableData;
        });
    }
    generate(tableData) {
        const dialect = dialects_1.dialects[this.sequelize.getDialect()];
        const generator = new auto_generator_1.AutoGenerator(tableData, dialect, this.options);
        return generator.generateText();
    }
    write(tableText) {
        const writer = new auto_writer_1.AutoWriter(tableText, this.options);
        return writer.write();
    }
    getDefaultPort(dialect) {
        switch (dialect) {
            case 'mssql':
                return 1433;
            case 'postgres':
                return 5432;
            default:
                return 3306;
        }
    }
}
exports.SequelizeAuto = SequelizeAuto;
module.exports = SequelizeAuto;
module.exports.SequelizeAuto = SequelizeAuto;
module.exports.default = SequelizeAuto;
